### DISCUSSION TITLE
No tags file?

### DISCUSSION THREAD

**Discussion by zephod77** [mrcjkb/rustaceanvim#307] [Q&A]
I set up rustaceanvim with as basic a config as I could and it seemed to work fine. I get warnings, and hints and code completion works. 
Next I wanted to add some key bindings but I realized I didn't know what commands to bind keys to, so I found this [rust-analyzer link](https://blog.logrocket.com/intro-to-rust-analyzer/) and found that <C-]> should jump to the definition of the thing under the cursor but instead I get "Error: E433: No tags file". Clearly, I'm missing something. I did some googling on tags and found talk of exuberant-tags and similar but that I don't have that installed on my Debian based system.
Interestingly, to me at least, <C-]> works in a C++ file.
Can someone tell me what config options I'm missing?

**Comment by mrcjkb** [comment]
Hey 👋 

Tags are not an LSP feature. You need a specialised tool like [rusty-tags](https://github.com/dan-t/rusty-tags) to generate them.
[Neovim uses `<C-]>` to first call `vim.lsp.buf.definition()` and then fall back to searching for tags if the LSP server doesn't respond with a definition to jump up](https://neovim.io/doc/user/lsp.html#vim.lsp.tagfunc()).

If you don't want to see that error message when the is no tag file and no LSP definition, set up a keymap for `vim.lsp.buf.definition()` 

See also the tip box in the [quick setup section](https://github.com/mrcjkb/rustaceanvim?tab=readme-ov-file#quick-setup) of this plugin's readme.

**Reply by zephod77** [reply]
OK, thanks. I installed rusty-tags and exuberant-ctags, run 'rusty-tags vi' in the project directory, open the .rs file, wait a while... wait a while longer, and then it works. It took about 10s for a 10 line file.

I find it strange that in all the videos on configuring nvim that I've seen, and I seen a lot of them, I don't recall anyone mentioning rusty-tags or that you have to generate the tags before you can do this sort of navigation. They just type gd and it works. Huh? I just typed gd and it did the same thing as <C-]> but I don't have any keymap defined. Where is that coming from? Sorry, Not your problem.

I do have a question for you though. Can I use the rustaceanvim plugin config to do this:
```
vim.g.rustaceanvim = {
    tools = {},
    server = {
        on_attach = function(client, bufnr)
            -- cmd to run rusty-tags vi in the project dir of the file just opened
        end
    },
    dap = {},
}```
Happy Friday!

**Reply by mrcjkb** [reply]
> I find it strange that in all the videos on configuring nvim that I've seen, and I seen a lot of them, I don't recall anyone mentioning rusty-tags

That's because tags are a less known feature these days.

> They just type gd and it works. 

Most people have a keymap set for that. If you're using a neovim distribution, it likely does too.

> I do have a question for you though. Can I use the rustaceanvim plugin config to do this

I have an issue open to add support for rusty-tags (if it's not too much of a performance impact, although I find rust-analyzer quite sufficient in most cases).

I've already [implemented it for fast-tags in haskell-tools.nvim](https://github.com/mrcjkb/haskell-tools.nvim/blob/master/lua/haskell-tools/tags.lua). You could use that as a reference.
